<?xml version="1.0"?>

<robot xmlns:sensor="http://playerstage.sourceforge.net/gazebo/xmlschema/#sensor"
	   xmlns:controller="http://playerstage.sourceforge.net/gazebo/
	   xmlschema/#controller"
	   xmlns:interface="http://playerstage.sourceforge.net/gazebo/xmlschema/#interface"
	   xmlns:xacro="http://ros.org/wiki/xacro"
	   >

	<xacro:include filename="$(find amr_robots_description)/urdf/pioneer3dx_wheel.xacro"/>

	<!-- Chassis description -->
	<link name="base_link">
		<!-- CHECK THESE VALUES. THEY MAY BE COMPLETELY WRONG. -->
		<inertial>
			<mass value="40.0"/>
			<origin xyz="-0.045 0 0.148"/>
			<inertia ixx="0.3338" ixy="0.0" ixz="0.0"
				 iyy="0.4783" iyz="0.0"
				 izz="0.3338"/>
		</inertial>
		<!-- The base visual is fine. Just the inertial is questionable.-->
		<visual name="pioneer_geom">
			<origin xyz="-0.045 0 0.148" rpy="0 0 0"/>
			<geometry name="pioneer_geom">
				<mesh filename="package://amr_robots_description/meshes/p3dx_meshes/chassis.stl"/>
			</geometry>
			<material name="ChassisRed">
				<color rgba="0.851 0.0 0.0 1.0"/>
			</material>
		</visual>
		<collision>
			<origin xyz="-0.045 0 0.148" rpy="0 0 0"/>
				<geometry>
					<mesh filename="package://amr_robots_description/meshes/p3at_meshes/chassis.stl"/>
				</geometry>
		</collision>
	</link>
	<gazebo reference="base_link">
		<material value="Gazebo/Red"/>
	</gazebo>

	<link name="top_plate">
		<inertial>
			<mass value="0.01"/>	
			<origin xyz="0 0 0"/>
			<inertia ixx="1" ixy="0" ixz="0"
				 iyy="1" iyz="0"
 				izz="1"/>
		</inertial>
		<visual name="pioneer_geom">
			<origin xyz="0 0 0" rpy="0 0 0"/>
			<geometry name="pioneer_geom">
				<mesh filename="package://amr_robots_description/meshes/p3dx_meshes/top.stl"/>
			</geometry>
			<material name="TopBlack">
					<color rgba="0.038 0.038 0.038 1.0"/>
			</material>
		</visual>
	</link>
	<gazebo reference="top_plate">
		<material value="Gazebo/Black"/>
	</gazebo>

	<joint name="base_top_joint" type="fixed">
		<origin xyz="-0.045 0 0.24" rpy="0 0 0"/>
		<parent link="base_link"/>
		<child link="top_plate"/>
	</joint>

	<!-- Swivel -->
	<joint name="base_swivel_joint" type="continuous">
		<origin xyz="-0.185 0 0.055" rpy="0 0 0"/>
		<anchor xyz="0 0 0"/>
    	<limit effort="100" velocity="100" k_velocity="0" />
    	<joint_properties damping="0.0" friction="0.0" />
		<parent link="base_link"/>
		<child link="swivel"/>
	</joint>
	<link name="swivel">
		<inertial>
			<mass value="0.1"/>	
			<origin xyz="0 0 0"/>
			<inertia ixx="0.01" ixy="0" ixz="0"
				 iyy="0.01" iyz="0" izz="0.01"/>
  	 	</inertial>
	<visual name="pioneer_geom">
		<origin xyz="0 0 0" rpy="0 0 0"/>
		<geometry name="pioneer_geom">
			<mesh filename="package://amr_robots_description/meshes/p3dx_meshes/swivel.stl"/>
		</geometry>
		<material name="swivel">
			<color rgba="0.5 0.5 0.5 1"/>
		</material>
	</visual>
	<collision>
		<origin xyz="0 0 0" rpy="0 0 0"/>
		<geometry>
			<box size="0.1 0.1 0.1"/>
		</geometry>
	</collision>
	</link>
	<gazebo reference="swivel">
		<material value="Gazebo/Grey"/>
	</gazebo>

<!-- Center Wheel + Hubcap -->
	<link name="center_hubcap">
		<inertial>
			<mass value="0.01"/>	
			<origin xyz="0 0 0"/>
			<inertia ixx="0.012411765597" ixy="-0.000711733678" ixz="0.00050272983"
				 iyy="0.015218160428" iyz="-0.000004273467" izz="0.011763977943"/>
  	 	</inertial>
	<visual name="pioneer_geom">
		<origin xyz="0 0 0" rpy="0 0 0"/>
		<geometry name="pioneer_geom">
			<mesh filename="package://amr_robots_description/meshes/p3dx_meshes/center_hubcap.stl"/>
		</geometry>
		<material name="swivel">
			<color rgba="0.5 0.5 0.5 1"/>
		</material>
	</visual>
	<collision>
		<origin xyz="0 0 0" rpy="0 0 0"/>
		<geometry>
			<box size="0.05 0.05 0.05"/>
		</geometry>
	</collision>
	</link>
	<gazebo reference="center_hubcap">
		<material value="Gazebo/Grey"/>
	</gazebo>

	<joint name="swivel_hubcap_joint" type="continuous">
		<origin xyz="-0.026 0 -0.016" rpy="0 0 0"/>
		<axis xyz="0 1 0"/>
		<anchor xyz="0 0 0"/>
    <limit effort="100" velocity="100" k_velocity="0" />
    <joint_properties damping="0.0" friction="0.0" />
	  <parent link="swivel"/>
		<child link="center_wheel"/>
	</joint>

	<link name="center_wheel">
		<inertial>
			<mass value="0.1"/>	
			<origin xyz="0 0 0"/>
			<inertia ixx="0.012411765597" ixy="-0.000711733678" ixz="0.00050272983"
				 iyy="0.015218160428" iyz="-0.000004273467" izz="0.011763977943"/>
  	 	</inertial>
	<visual name="pioneer_geom">
		<origin xyz="0 0 0" rpy="0 0 0"/>
		<geometry name="pioneer_geom">
			<mesh filename="package://amr_robots_description/meshes/p3dx_meshes/center_wheel.stl"/>
		</geometry>
		<material name="WheelBlack">
			<color rgba="0.117 0.117 0.117 1"/>
		</material>
	</visual>
	<collision>
		<origin xyz="0 0 0" rpy="${-3.1415927/2.0} 0 0"/>
		<geometry>
			<!--<mesh filename="package://amr_robots_description/meshes/p3dx_meshes/center_wheel.stl"/>-->
			<cylinder radius="0.0375" length="0.01"/>
		</geometry>
	</collision>
	</link>
	<gazebo reference="left_hub">
		<material value="Gazebo/Yellow"/>
	</gazebo>

	<joint name="center_wheel_joint" type="fixed">
		<origin xyz="-0.0035 0 -0.001" rpy="0 0 0"/>
		<parent link="center_wheel"/>
		<child link="center_hubcap"/>
	</joint>

	<link name="left_hub">
		<visual>
			<origin xyz="0 0 0" rpy="0 0 0"/>
			<geometry>
				<mesh filename="package://amr_robots_description/meshes/p3dx_meshes/left_hubcap.stl"/>
			</geometry>
			<material name="HubcapYellow">
				<color rgba="1.0 0.811 0.151 1.0"/>
			</material>
		</visual>
		<collision>
			<origin xyz="0 0 0" rpy="${-3.1415927/2.0} 0 0"/>
			<geometry>
				<!--<mesh filename="package://amr_robots_description/meshes/p3dx_meshes/center_wheel.stl"/>-->
				<cylinder radius="0.09" length="0.01"/>
			</geometry>
		</collision>
	</link>

	<joint name="left_hub_joint" type="fixed">
		<origin xyz="0 0.15 0.08" rpy="0 0 0"/>
		<parent link="base_link"/>
		<child link="left_hub"/>
		<axis xyz="0 1 0"/>
	</joint>

	<link name="left_wheel_joint">
		<visual>
			<origin xyz="0 0 0" rpy="0 0 0"/>
			<geometry>
				<mesh filename="package://amr_robots_description/meshes/p3dx_meshes/left_wheel.stl"/>
			</geometry>
			<material name="WheelBlack">
				<color rgba="0.117 0.117 0.117 1"/>
			</material>
		</visual>
	</link>

	<joint name="left_wheel_joint" type="fixed">
		<origin xyz="0 0 0" rpy="0 0 0"/>
		<parent link="left_hub"/>
		<child link="left_wheel_joint"/>
	</joint>
	
	<gazebo reference="left_wheel_joint">
		<material value="Gazebo/Black"/>
	</gazebo>

	<link name="right_hub">
		<visual>
			<origin xyz="0 0 0" rpy="0 0 0"/>
			<geometry>
				<mesh filename="package://amr_robots_description/meshes/p3dx_meshes/right_hubcap.stl"/>
			</geometry>
			<material name="HubcapYellow">
				<color rgba="1.0 0.811 0.151 1.0"/>
			</material>
		</visual>
		<collision>
		<origin xyz="0 0 0" rpy="${-3.1415927/2.0} 0 0"/>
		<geometry>
			<!--<mesh filename="package://amr_robots_description/meshes/p3dx_meshes/center_wheel.stl"/>-->
			<cylinder radius="0.09" length="0.01"/>
		</geometry>
	</collision>
	</link>
	<gazebo reference="right_hub">
		<material value="Gazebo/Yellow"/>
	</gazebo>

	<joint name="right_hub_joint" type="fixed">
		<origin xyz="0 -0.15 0.08" rpy="0 0 0"/>
		<parent link="base_link"/>
		<child link="right_hub"/>
		<axis xyz="0 1 0"/>
	</joint>

	<link name="right_wheel_joint">
		<visual>
			<origin xyz="0 0 0" rpy="0 0 0"/>
			<geometry>
				<mesh filename="package://amr_robots_description/meshes/p3dx_meshes/right_wheel.stl"/>
			</geometry>
			<material name="WheelBlack">
				<color rgba="0.117 0.117 0.117 1"/>
			</material>
		</visual>
	</link>

	<joint name="right_wheel_joint" type="fixed">
		<origin xyz="0 0 0" rpy="0 0 0"/>
		<parent link="right_hub"/>
		<child link="right_wheel_joint"/>
	</joint>
	<gazebo reference="right_wheel_joint">
		<material value="Gazebo/Black"/>
	</gazebo>

	<!-- Front + Back Sonar -->
	<link name="front_sonar">
		<inertial>
			<mass value="0.0001"/>	
			<origin xyz="0 0 0"/>
			<inertia ixx="1" ixy="0" ixz="0"
				 iyy="1" iyz="0" izz="1"/>
		</inertial>
		<visual name="pioneer_geom">
			<origin xyz="0 0 0" rpy="0 0 0"/>
			<geometry name="pioneer_geom">
				<mesh filename="package://amr_robots_description/meshes/p3dx_meshes/front_sonar.stl"/>
			</geometry>
			<material name="SonarYellow">
				<color rgba="0.715 0.583 0.210 1.0"/>
			</material>
		</visual>
		<collision>
			<origin xyz="0 0 0" rpy="0 0 0"/>
			<geometry>
				<box size="0.1 0.1 0.1"/>
			</geometry>
		</collision>
	</link>

	<gazebo reference="front_sonar">
		<material value="Gazebo/Yellow"/>
	</gazebo>

	<joint name="base_front_joint" type="fixed">
		<origin xyz="-0.198 0 0.208" rpy="0 0 0"/>
		<parent link="base_link"/>
		<child link="front_sonar"/>
	</joint>

	<link name="back_sonar">
		<inertial>
			<mass value="0.0001"/>	
			<origin xyz="0 0 0"/>
			<inertia ixx="1" ixy="0" ixz="0"
				 iyy="1" iyz="0" izz="1"/>
		</inertial>
		<visual name="back_sonar_vis">
			<origin xyz="0 0 0" rpy="0 0 0"/>
			<geometry name="pioneer_geom">
				<mesh filename="package://amr_robots_description/meshes/p3dx_meshes/back_sonar.stl"/>
			</geometry>
			<material name="SonarYellow">
				<color rgba="0.715 0.583 0.210 1.0"/>
			</material>
		</visual>
		<collision>
			<origin xyz="0 0 0" rpy="0 0 0"/>
			<geometry>
				<box size=" 0.5 0.25 0.05"/>
			</geometry>
		</collision>
	</link>

	<gazebo reference="back_sonar">
		<material value="Gazebo/Yellow"/>
	</gazebo>

	<joint name="base_back_joint" type="fixed">
		<origin xyz="0.109 0 0.209" rpy="0 0 0"/>
		<!-- <origin xyz="0 0 0" rpy="0 0 0"/> -->
		<parent link="base_link"/>
		<child link="back_sonar"/>
	</joint>

	<xacro:macro name="p3dx_gazebo_ext">

	<gazebo>
	  <plugin name="skid_steer_drive_controller" filename="libgazebo_ros_skid_steer_drive.so">
	    <updateRate>100.0</updateRate>
	    <robotNamespace>/</robotNamespace>
	    <leftFrontJoint>base_swivel_joint</leftFrontJoint>
	    <rightFrontJoint>base_swivel_joint</rightFrontJoint>
	    <leftRearJoint>swivel_hubcap_joint</leftRearJoint>
	    <rightRearJoint>swivel_hubcap_joint</rightRearJoint>
	    <wheelSeparation>0.3</wheelSeparation>
	    <wheelDiameter>0.1</wheelDiameter>
	    <robotBaseFrame>base_link</robotBaseFrame>
	    <torque>200</torque>
	    <commandTopic>cmd_vel</commandTopic>
	    <odometryTopic>odom</odometryTopic>
	    <odometryFrame>odom</odometryFrame>
	  </plugin>
	</gazebo>

	</xacro:macro>
</robot>
